handlers.go

```diff
@@ -3526,3 +3528,52 @@ func (s *Server) handleGetPublicNodeInfo(c echo.Context) error {
 		PrimaryAddress: s.FilClient.ClientAddr,
 	})
 }
+
+type retrievalCandidate struct {
+	Miner   address.Address
+	RootCid cid.Cid
+	DealID  uint
+}
+
+func (s *Server) handleGetRetrievalCandidates(c echo.Context) error {
+	// Read the cid from the client request
+	cid, err := cid.Decode(c.Param("cid"))
+	if err != nil {
+		return c.JSON(http.StatusBadRequest, map[string]string{
+			"error": "invalid cid",
+		})
+	}
+
+	var candidateInfos []struct {
+		miner  string
+		cid    util.DbCID
+		dealID uint
+	}
+	if err := s.DB.
+		Table("content_deals").
+		Where("content IN ? AND NOT failed",
```

Comment:
youll have to wrap the `?` in parentheses for subqueries, gorm doesnt do it on its own for some reason